ui/app/scripts/stream/services/graph-to-text.js

```diff
@@ -0,0 +1,417 @@
+		linksToVisit = {};
+		nodesToVisit = {};
+		nodesIncomingLinksCount = {};
+		g = graph;
+		g.getElements().forEach(function(element) {
+			if (isNode(element)) {
+				var nodeId = element.get('id');
+				var incomingLinkCount = 0;
+				g.getConnectedLinks(element, {inbound: true}).forEach(function(link) {
+					if (linkSourceIsNode(link,g)) {
+						linksToVisit[link.get('id')] = link;
+						numberOfLinksToVisit++;
+						incomingLinkCount++;
+					}
+				});				
+				nodesToVisit[nodeId] = element;
+				numberOfNodesToVisit++;
+				nodesIncomingLinksCount[nodeId] = incomingLinkCount;
+			}
+		});
+	}
+    
+    function isTapLink(link) {
+        return link.get('source').port === 'tap';
+    }
+    
+	function printStream(stream) {
+		var text = 'Stream: ';
+		for (var i=0;i<stream.length;i++) {
+			if (i>0) {
+				text+=' ';
+			}
+			text+=getName(stream[i]);
+		}
+		console.log(text);
+	}
+
+	function getIncomingLinks(node) {
+		// Only possible if call is occurring during link drawing (one end connected but not the other)
+		if (!node) {return [];}
+		return g.getConnectedLinks(node, {inbound: true});
+	}
+
+	function getOutgoingStreamLinks(node) {
+		var links = [];
+		// !node only possible if call is occurring during link drawing (one end connected but not the other)
+		if (node) {
+			g.getConnectedLinks(node, {outbound: true}).forEach(function(link) {
+				var source = link.get('source');
+				if (source.port === 'output') {
+					links.push(link);
+				}
+			});
+		}
+		return links;
+	}
+
+	function getOutgoingTapLinks(node) {
+		var links = [];
+		// !node only possible if call is occurring during link drawing (one end connected but not the other)
+		if (node) {
+			g.getConnectedLinks(node, {outbound: true}).forEach(function(link) {
+				var source = link.get('source');
+				if (source.port === 'tap') {
+					links.push(link);
+				}
+			});
+		}
+		return links;
+	}
+
+	function findStreamName(node) {
+		var streamName = node.attr('stream-name');
+		if (streamName) {
+			return streamName;
+		}
+		// Go up the link chain to find the named element
+		var incomingLinks = getIncomingLinks(node);
+		if (incomingLinks && incomingLinks.length > 0 && !(incomingLinks.length === 1 && incomingLinks[0].get('source').port === 'tap')) {
+			if (incomingLinks.length > 1) {
+				throw {'msg': 'nodes should only have 1 incoming link at most. Node '+getName(node)+' has '+incomingLinks.length};
+			}
+			var sourceId = incomingLinks[0].get('source').id;
+			var source = g.getCell(sourceId);
+			return findStreamName(source);
+		}
+		return 'UNKNOWN';
+	}
+
+	// Create a destination of the form :streamname.name
+	// The 'name' element will be the node name unless a label is provided. If a label is
+	// provided it will be used instead.
+	function toTapDestination(node) {
+		var appname = node.attr('node-name');
+		if (!appname) {
+			appname = node.attr('metadata/name');
+		}
+		return ':'+findStreamName(node)+'.'+appname;
```

Comment:
It would be better if we move the tap prefix `colon` to a `var` so that it can be reused in all other references for tap prefix.